source/ceylon/logging/LoggerFactory.ceylon

```diff
@@ -0,0 +1,90 @@
+import ceylon.language.meta.model {
+    ClassOrInterface
+}
+import ceylon.collection {
+    MutableMap,
+    HashMap
+}
+
+"`LoggerFactory` can be satisfied to define a custom
+ logging strategy. Typically, an instance of `LoggerFactory`
+ will be be assigned to [[loggerFactory]]."
+shared
+interface LoggerFactory {
+
+    "Function to obtain a [[Logger]] for the given
+     [[Category]].
+
+     Callers intending to wrap the returned logger should
+     provide the [[ClassOrInterface]] of the wrapper.
+     For example:
+
+            FormattingLogger formattingLogger(Category category)
+                =>  FormattingLogger(loggerFactory.logger(
+                        category, `FormattingLogger`));"
+    shared formal
+    Logger logger<Wrapper>(
+            Category category,
+            "The class that will be used to wrap the returned
+             value, if any. This may be used by loggers that
+             log source file and line number information."
+            ClassOrInterface<Wrapper>? wrapper=null)
+            given Wrapper satisfies Object;
+}
```

Comment:
Unless I can get you to buy into some future proofing with an `interface LoggerFactory`, I propose we simply change the signature of `Logger.log()` to:

```ceylon
void log(
        Priority priority,
        Message message,
        Throwable? throwable=null,
        ClassDeclaration? wrapper=null);
```

with a dependency on #368. This would have the added benefit of allowing implementations to cache loggers and ensure a 1-to-1 mapping of logger-instance-to-category, if for some reason that's important.

There are a multitude of options - I can replace this commit with whatever you decide.